packages/back-end/src/queryRunners/QueryRunner.ts

```diff
@@ -443,6 +454,34 @@ export abstract class QueryRunner<
     }
   }
 
+  public async executeQueryWhenReady<
+    Rows extends RowsType,
+    ProcessedRows extends ProcessedRowsType
+  >(
+    doc: QueryInterface,
+    run: (
+      query: string,
+      setExternalId: ExternalIdCallback
+    ) => Promise<QueryResponse<Rows>>,
+    process: (rows: Rows) => ProcessedRows,
+    currentTimeout: number = 500
+  ): Promise<void> {
+    // If too many queries are running against the datastore, use capped exponential backoff to wait until they've finished
+    const concurrencyLimitReached = await this.concurrencyLimitReached();
+    if (concurrencyLimitReached) {
+      logger.debug(
+        `${doc.id}: Query concurrency limit reached, waiting ${currentTimeout} before retrying`
+      );
+      const nextTimeout = Math.min(currentTimeout * 2, 8000);
+      // TODO: ran into issue with orphaned queries here; is that dev-only or do we need to check for stale queries
```

Comment:
Much more common on dev since Node is constantly being killed and restarted, but can still happen on production, especially during deploys.  Stale queries should be cleaned up within 2 minutes max.  That's because we record a heart beat every 30 seconds and if a query misses 2 heart beats, we assume it's orphaned and kill it.  So 60 seconds + cron every minute = up to 2 minutes stale.